libvast/src/system/spawn.cpp

```diff
@@ -11,91 +11,142 @@
-  using namespace vast::binary_byte_literals;
-  auto mss = size_t{128};
-  auto segments = size_t{10};
-  auto r = opts.params.extract_opts({
-    {"segments,s", "number of cached segments", segments},
-    {"max-segment-size,m", "maximum segment size in MB", mss}
-  });
-  opts.params = r.remainder;
-  if (!r.error.empty())
-    return make_error(ec::syntax_error, r.error);
-  mss *= 1_MiB;
-  auto a = self->spawn(archive, opts.dir / opts.label, segments, mss);
-  return actor_cast<actor>(a);
-}
-
-expected<actor> spawn_exporter(node_actor* self, options& opts) {
-  auto max_events = uint64_t{0};
-  auto r = opts.params.extract_opts({
-    {"continuous,c", "marks a query as continuous"},
-    {"historical,h", "marks a query as historical"},
-    {"unified,u", "marks a query as unified"},
-    {"events,e", "maximum number of results", max_events},
-  }, nullptr, true);
-  if (!r.error.empty())
-    return make_error(ec::syntax_error, r.error);
-  if (r.remainder.empty())
+using namespace vast::binary_byte_literals;
+
+// Convenience alias for the system defaults namespace.
+namespace sd = vast::defaults::system;
+
+// Convenience alias for the command defaults namespace.
+namespace cd = vast::defaults::command;
+
+namespace vast::system {
+
+namespace {
+
+// Attempts to parse [args.first, args.last) as vast::expression and
+// returns a normalized and validated version of that expression on success.
+caf::expected<expression> normalized_and_valided(spawn_arguments& args) {
+  if (args.empty())
     return make_error(ec::syntax_error, "no query expression given");
-  auto str = r.remainder.get_as<std::string>(0);
-  for (auto i = 1u; i < r.remainder.size(); ++i)
-    str += ' ' + r.remainder.get_as<std::string>(i);
-  // Parse expression.
-  auto expr = to<expression>(str);
-  if (!expr)
-    return expr.error();
-  expr = normalize_and_validate(*expr);
-  if (!expr)
-    return expr.error();
+  if (auto e = to<expression>(caf::join(args.first, args.last, " ")); !e)
+    return std::move(e.error());
+  else
+    return normalize_and_validate(*e);
+}
+
+// Attemps to read a schema file and parse its content. Can either 1) return
+// nothing if the user didn't specifiy a schema file, 2) produce a valid
+// schema, or 3) run into an error.
+caf::expected<caf::optional<schema>> read_schema(spawn_arguments& args) {
+  auto schema_file_ptr = get_if<std::string>(&args.options, "global.schema");
+  if (!schema_file_ptr)
+    return caf::optional<schema>{caf::none};
+  if (auto str = load_contents(*schema_file_ptr); !str)
+    return std::move(str.error());
+  else if (auto result = to<schema>(*str); !result)
+    return std::move(result.error());
+  else
+    return caf::optional<schema>{std::move(*result)};
+}
+
+} // namespace <anonymous>
+
+maybe_actor spawn_archive(local_actor* self, spawn_arguments& args) {
+  if (!args.empty())
+    return make_error(ec::syntax_error, "unexpected argument(s)");
```

Comment:
It would add a lot of value if the received arguments were inserted in the error message. I propose to factor this into a function:
```c++
caf::error unexpected_arguments(const spawn_arguments& args) {
  return make_error(ec::syntax_error, "unexpected argument(s)",
                    caf::join(args.first, args.last, " "));
}

```